nixos/modules/services/editors/emacs.nix

```diff
@@ -0,0 +1,74 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+
+  cfg = config.services.emacs;
+
+  editorScript = pkgs.writeScriptBin "editor-emacs.sh" ''
+    #!${pkgs.stdenv.shell}
+    if [ -z "$1" ]; then
+      exec ${cfg.package}/bin/emacsclient --create-frame --alternate-editor ${cfg.package}/bin/emacs
+    else
+      exec ${cfg.package}/bin/emacsclient --alternate-editor ${cfg.package}/bin/emacs "$@"
+    fi
+  '';
+
+in {
+
+  options.services.emacs = {
+    enable = mkOption {
+      type = types.bool;
+      default = false;
+      example = true;
+      description = ''
+        Whether to install a user service for the Emacs daemon. Once
+        the service is started, use emacsclient to connect to the
+        daemon.
+
+        The service must be manually started and/or enabled for each
+        user with "systemctl --user start emacs" and
+        "systemctl --user enable emacs".
+      '';
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.emacs;
+      defaultText = "pkgs.emacs";
+      description = ''
+        emacs derivation to use.
+      '';
+    };
+
+    defaultEditor = mkOption {
+      type = types.bool;
+      default = false;
+      example = true;
+      description = ''
+        When enabled, configures emacsclient to be the default editor
+        using the EDITOR environment variable.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.user.services.emacs = {
+      description = "Emacs: the extensible, self-documenting text editor";
+
+      serviceConfig = {
+        Type      = "forking";
+        ExecStart = "${pkgs.bash}/bin/bash -c 'source ${config.system.build.setEnvironment}; exec ${cfg.package}/bin/emacs --daemon'";
```

Comment:
Hey! That’s a great idea. However, I think probably almost everyone would want their custom environment variables visible in Emacs.

Maybe we could add an option to have this? One solution would be to run:

`${pkgs.bash}/bin/bash -c 'source ${config.system.build.setEnvironment} ; exec $SHELL --login -c "exec ${cfg.package}/bin/emacs --daemon"'`

… and the `--login` part would be optional and configurable with some boolean.

I’m doing exactly this here: https://github.com/michalrus/dotfiles/blob/3207a86b44ed9f2dd5de7f803bedc078f8754876/.config/systemd/user/emacs-daemon.start.symlink (I’m also removing `.emacs.desktop.lock` — a leftover from unclean exits, but that’s not that important).